Libraries/LibWeb/CSS/SelectorEngine.cpp

```diff
@@ -759,6 +760,140 @@ static inline bool matches_pseudo_class(CSS::Selector::SimpleSelector::PseudoCla
 
         return false;
     }
+    case CSS::PseudoClass::Valid: {
+        // https://html.spec.whatwg.org/#selector-valid
+        // The :valid pseudo-class must match any element falling into one of the following categories:
+
+        // - elements that are candidates for constraint validation and that satisfy their constraints
+        if (is<Web::HTML::FormAssociatedElement>(element)) {
+            auto& form_associated_element = as<Web::HTML::FormAssociatedElement>(element);
+            if (form_associated_element.is_candidate_for_constraint_validation() && form_associated_element.satisfies_its_constraints())
+                return true;
+        }
+
+        // - form elements that are not the form owner of any elements that themselves are candidates for constraint validation but do not satisfy their constraints
+        if (is<Web::HTML::HTMLFormElement>(element)) {
+            auto& form_element = as<Web::HTML::HTMLFormElement>(element);
+            bool has_invalid_elements = false;
+            element.for_each_in_subtree([&](auto& node) {
+                if (auto const* form_associated_element = dynamic_cast<Web::HTML::FormAssociatedElement const*>(&node)) {
```

Comment:
Nit: We've recently started preferring `as_if<T>()` to `dynamic_cast`. (This didn't work until a couple of days ago, in case you tried that before.)

Reply:
Done.
And no, I hadn't tried that yet, I just took this way of doing this from elsewhere in the code.